CSCI-SHU-410-SE-Project/Deskulpt
# FEAT Frontend wrapper for `@deskulpt/apis`

In this PR, I have implemented the frontend wrapper for default dependencies. Default dependencies are dependencies provided to widgets by Deskulpt. The wrapper covers two types of default dependencies
- those that are shared by all widgets, such as `React.useState()`
- those that are widget-specific, such as `readFile()`
 
Below is a brief example of widget source code that uses these two types of dependencies. For a full example, see [this writer widget](https://github.com/CSCI-SHU-410-SE-Project/test-widgets/tree/main/ok-writer)
```typescript
import React from "@deskulpt/react"
import apis from "@deskulpt/apis"

console.log(React.useState)
console.log(apis.fs.readFile)
```

## Default depedencies shared by all widgets

These are dependencies shared by all widgets (and by Deskulpt manager). For now, I only wrap the default export `React` of react. A widget may use the react from default dependencies like this
```typescript
import React from "@deskulpt/react"

console.log(React.useState)
```
This is achieved by several level of indirections.

First, we have a wrapper of react in Deskulpt source code
```typescript
// src/@deskulpt/react.index.ts
import React from "react";
export default React;
```

To make this wrapper module accessible to the webpage in canvas window (and thus to all widgets in that window) when Deskulpt is running, we do two things

1. When building Deskulpt, we let rollup preserve module structure by setting `preserveModule` to true in vite config's rollup options. This way, rollup won't bundle the source code. Instead, it will only convert node module import to relative path import, and file structure won't be changed after building [^1]. The problem with bundling is that variables and functions will be renamed and module structure will change. For example, rollup may convert `export default React` in our source code to `export { React as R }` in the bundle.

	When Deskulpt is running, the source code for our wrapper of react is configured to be stored at `/assets/src/@deskulpt/react/index.js`, and js code running in the browser can import from it via `import React from "/assets/src/@deskulpt/react/index.js"`

2. To redirect import from `@deskulpt/react` to `/assets/src/@deskulpt/react/index.js`, we create an importmap in the canvas webpage [^2]
   ```json
	{
	"imports": {
		"@deskulpt/react": "/assets/src/@deskulpt/react/index.js",
		}
	}
   ```

## Default dependencies that are widget-specific

In Deskulpt, these types of deps are mainly widget apis. For example, consider the widget api `async function readFile(path: string): string` that reads a file from the widget's folder and returns the content as string. A call of `readFile("output.txt")` by different widgets should read from `output.txt` in different widget folders. 

Roughly speaking, this is implemented by letting different widgets import different `readFile` that have different `widgetId` "associated" to it. 

### First Attempt

Consider two widgets with ids `todolist` and `calendar`. A first attempt would be to let widget developer explicitly import from different default dependencies.

For `todolist`, we have
```typescript
// $APPDATA/widgets/todolist/src/App.tsx

import React from "@deskulpt/react"
import apis from "todolist/@deskulpt/apis"

const content = await apis.fs.readFile("output.txt")
// Equivalent to 
// window.__TAURI_INVOKE__({
//	cmd: "plugin:widget_api.fs|read_file", 
//  widgetId: "todolist", 
//  path: "output.txt"
// })
```

For `calendar`, we have
```typescript
// $APPDATA/widgets/calendar/src/App.tsx

import React from "@deskulpt/react"
import apis from "calendar/@deskulpt/apis"

const content = await apis.fs.readFile("output.txt")
// Equivalent to 
// window.__TAURI_INVOKE__({
//	cmd: "plugin:widget_api.fs|read_file", 
//  widgetId: "calendar", 
//  path: "output.txt"
// })
// Note that the widgetId in the argument is different
```

The `todolist/@deskulpt/apis` should points to a javascript file like this (similar for `calendar`)
```javascript
// Source code of `todolist/@deskulpt/apis` module
// This is file is created at the runtime of Deskulpt so that it can 
// dynamically adapts to different widget ids

// Import from the raw apis that require explicitly passing in widgetId
import * as rawApis from "@deskulpt/apis"

let apis = {}
// Create a readFile function that calls the raw readFile api and pass in the widgetId
apis["fs"]["readFile"] = (...args) => rawApis.fs.readFile("todolist", ...args)
// ... wrapper for other raw widget apis

exports default apis;
```

This will work, but it has a catch: the widget developer need to explicitly write out the widget id in their source code. This is undesirable because it exposes too much details to widget developers, making the widget development error-prone (e.g. a widget dev may misspell the widget id). The widget id is the name of the widget folder in `$APPDATA/widgets`, and it should be managed by the Deskulpt manager.

### Solution: Redirection of import during widget bundling

What we want is that different widgets could import different apis by the same line of code below

```typescript
import apis from "@deskulpt/apis"
```

This is possible because we are in full control of the widget compilation. We can create the widget-specific import module and swap it in during widget compilation. This is achieved in two steps

1. Right before Deskulpt manager (javascript) invoke `bundle_widget`, we create a blob object of javascript code of the `todolist/@deskulpt/apis` module, and pass the url of the blob as an argument to `bundle_widget`.
2. In Rust backend, the bundler locates import statements whose source is `"@deskulpt/apis"`, and swaps it with the blob url.

This way, the widget developer doesn't need to explicitly specify the widget id in their source code. Instead, they can import from the same `@deskulpt/apis` module for different widgets, and Deskulpt will convert it to imports of widget-specific apis.

## How to test this PR

Clone the PR, install npm package, and load the [test widget](https://github.com/CSCI-SHU-410-SE-Project/test-widgets/tree/main/ok-writer) to see if the wrapped react and widget apis work.


[^1]: Note that minification, tree-shaking and other steps won't be affected. The only problem with this approach is that there won't be a single js bundle but multiple small js files. This could slow down the initial loading time of the webpage in the canvas window because of the overhead of requesting multiple files, but the slowdown is near negligible since all files are local and requests are based on IPC.

[^2]: When debugging, the files won't be bundled, and the wrapper is located at `/src/@deskulpt/react/index.ts` instead of `/assets/src/@deskulpt/react/index.js`. Since importmap can't be dynamically updated when the webpage is loaded, we create two canvas webpage html `canvs-build.html` and `canvas-dev.html`, and let npm choose which one should be the `cavnas.html` (see the `predev` and `prebuild` npm commands)
